internal/component/pyroscope/relabel/relabel.go

```diff
@@ -0,0 +1,337 @@
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	defer c.exited.Store(true)
+	<-ctx.Done()
+	return nil
+}
+
+func (c *Component) Update(args component.Arguments) error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	newArgs := args.(Arguments)
+	newRCS := alloy_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelConfigs)
+
+	// If relabeling rules changed, purge the cache
+	if relabelingChanged(c.rcs, newRCS) {
+		level.Debug(c.opts.Logger).Log("msg", "received new relabel configs, purging cache")
+		c.cache.Purge()
+		c.metrics.cacheSize.Set(0)
+	}
+
+	if newArgs.MaxCacheSize != c.maxCacheSize {
+		evicted := c.cache.Resize(newArgs.MaxCacheSize)
+		if evicted > 0 {
+			level.Debug(c.opts.Logger).Log("msg", "resizing cache led to evicting items", "evicted_count", evicted)
+		}
+		c.maxCacheSize = newArgs.MaxCacheSize
+	}
+
+	c.rcs = newRCS
+	c.fanout.UpdateChildren(newArgs.ForwardTo)
+
+	c.opts.OnStateChange(Exports{
+		Receiver: c,
+		Rules:    newArgs.RelabelConfigs,
+	})
+
+	return nil
+}
+
+type ProfileDroppedError struct{}
+
+func (e *ProfileDroppedError) Error() string {
+	return "profile dropped by relabel rules"
+}
+
+func (c *Component) Append(ctx context.Context, lbls labels.Labels, samples []*pyroscope.RawSample) error {
+	if c.exited.Load() {
+		return fmt.Errorf("%s has exited", c.opts.ID)
+	}
+
+	c.mut.RLock()
+	defer c.mut.RUnlock()
+
+	newLabels, keep, err := c.relabel(lbls)
+	if err != nil {
+		return err
+	}
+	if !keep {
+		return &ProfileDroppedError{}
+	}
+
+	return c.fanout.Appender().Append(ctx, newLabels, samples)
+}
+
+func (c *Component) AppendIngest(ctx context.Context, profile *pyroscope.IncomingProfile) error {
+	if c.exited.Load() {
+		return fmt.Errorf("%s has exited", c.opts.ID)
+	}
+
+	c.mut.RLock()
+	defer c.mut.RUnlock()
+
+	lbls, err := extractLabelsFromIncomingProfile(profile)
+	if err != nil {
+		return err
+	}
+	if lbls.IsEmpty() {
+		return c.fanout.Appender().AppendIngest(ctx, profile)
+	}
+
+	newLabels, keep, err := c.relabel(lbls)
+	if err != nil {
+		return fmt.Errorf("processing labels: %w", err)
+	}
+	if !keep {
+		c.metrics.profilesDropped.Inc()
+		return &ProfileDroppedError{}
+	}
+
+	updateProfileLabels(profile, newLabels)
+	return c.fanout.Appender().AppendIngest(ctx, profile)
+}
+
+func (c *Component) Appender() pyroscope.Appender {
+	return c
+}
+
+type cacheItem struct {
+	original  model.LabelSet
+	relabeled model.LabelSet
+}
+
+// relabel applies the configured relabeling rules to the input labels
+// Returns the new labels, whether to keep the profile, and any error
+func (c *Component) relabel(lbls labels.Labels) (labels.Labels, bool, error) {
+	labelSet := toModelLabelSet(lbls)
+	hash := labelSet.Fingerprint()
+
+	// Check cache
+	if result, keep, found := c.getCacheEntry(hash, labelSet); found {
+		return result, keep, nil
+	}
+
+	// Apply relabeling
+	newLabels, keep := relabel.Process(lbls, c.rcs...)
```

Comment:
This probably needs a mircro benmark, but I think using the label builder can help reduce allocations:
https://pkg.go.dev/github.com/prometheus/prometheus/model/relabel#ProcessBuilder

Reply:
Good point, you're right:
```goos: darwin
goarch: arm64
pkg: github.com/grafana/alloy/internal/component/pyroscope/relabel
cpu: Apple M3 Pro
BenchmarkRelabelProcess/Process-11         	 3737449	       313.8 ns/op	     480 B/op	       8 allocs/op
BenchmarkRelabelProcess/ProcessBuilder-11  	 7209895	       164.1 ns/op	      80 B/op	       5 allocs/op
PASS
ok  	github.com/grafana/alloy/internal/component/pyroscope/relabel	3.332s
```